From 391af59ff74dff9bb5348a204dcb46284bd8359c Mon Sep 17 00:00:00 2001
From: Dong Chenchen <superdcc97@163.com>
Date: Mon, 27 Apr 2026 21:56:36 +0800
Subject: [PATCH 4/9] mm: convert VMA insert/remove to use per-shard locks

In sharded mode (i_mmap_shards_ptr != NULL), VMA insert and remove
operations use per-shard rw_semaphore locks instead of the global
i_mmap_rwsem. Each VMA is assigned to a shard based on its address
hash (vma_shard_idx), reducing lock contention when multiple CPUs
operate on VMAs mapping the same file.

Key changes:

- __vma_link_file/__remove_shared_vm_struct: Use per-shard tree and
  track vma_count in sharded mode
- vma_prepare/vma_complete: Lock/unlock relevant shard rwsems using
  i_mmap_lock_shards_write/i_mmap_unlock_shards_write
- vma_link_file: Lock per-shard rwsem in sharded mode
- unlink_file_vma_batch_process: Group VMAs by shard, lock shards in
  ascending index order to avoid deadlocks, then remove
- dup_mmap_file_batch:  Collect unique shard locks in sharded mode,
  use vma_interval_tree_insert (not _after) since parent VMA may be
  in a different shard
- i_mmap_lock_shards_write/i_mmap_unlock_shards_write: New helpers
  that collect unique shards from VMA array, sort by index, and
  acquire/release rwsems

Centralized mode behavior is unchanged.

Signed-off-by: GLM <glm@linux.dev>
---
 include/linux/fs.h |   6 +++
 mm/i_mmap_shards.c |  62 +++++++++++++++++++++
 mm/mmap.c          |  90 +++++++++++++++++++++++++++----
 mm/vma.c           | 132 +++++++++++++++++++++++++++++++++++----------
 mm/vma.h           |   5 ++
 5 files changed, 257 insertions(+), 38 deletions(-)

diff --git a/include/linux/fs.h b/include/linux/fs.h
index 51146ab1eb07..725235d9f48f 100644
--- a/include/linux/fs.h
+++ b/include/linux/fs.h
@@ -622,6 +622,12 @@ i_mmap_shard_by_idx(struct address_space *mapping, unsigned int idx)
 extern struct i_mmap_shards *i_mmap_shards_alloc(gfp_t gfp);
 extern void i_mmap_shards_free(struct i_mmap_shards *shards);
 extern int mapping_convert_to_shards(struct address_space *mapping);
+extern int i_mmap_lock_shards_write(struct address_space *mapping,
+				    struct i_mmap_shard *shards[],
+				    struct vm_area_struct *vmas[],
+				    int nr_vmas);
+extern void i_mmap_unlock_shards_write(struct i_mmap_shard *shards[],
+				       int nr_shards);
 
 /*
  * Might pages of this file have been modified in userspace?
diff --git a/mm/i_mmap_shards.c b/mm/i_mmap_shards.c
index f5add9003232..4e2bc23a5608 100644
--- a/mm/i_mmap_shards.c
+++ b/mm/i_mmap_shards.c
@@ -70,4 +70,66 @@ int mapping_convert_to_shards(struct address_space *mapping)
 	smp_store_release(&mapping->i_mmap_shards_ptr, shards);
 
 	return 0;
+}
+
+int i_mmap_lock_shards_write(struct address_space *mapping,
+			     struct i_mmap_shard *shards[],
+			     struct vm_area_struct *vmas[],
+			     int nr_vmas)
+{
+	struct i_mmap_shards *shards_ptr;
+	unsigned int shard_idx[I_MMAP_NR_SHARDS];
+	int nr_shards = 0;
+	int i, j;
+
+	shards_ptr = READ_ONCE(mapping->i_mmap_shards_ptr);
+	if (!shards_ptr)
+		return 0;
+
+	for (i = 0; i < nr_vmas; i++) {
+		unsigned int idx;
+
+		if (!vmas[i])
+			continue;
+
+		idx = vma_shard_idx(vmas[i]) % shards_ptr->nr_shards;
+
+		for (j = 0; j < nr_shards; j++) {
+			if (shard_idx[j] == idx)
+				break;
+		}
+		if (j < nr_shards)
+			continue;
+
+		shard_idx[nr_shards] = idx;
+		shards[nr_shards] = &shards_ptr->shards[idx];
+		nr_shards++;
+	}
+
+	for (i = 1; i < nr_shards; i++) {
+		unsigned int tmp_idx = shard_idx[i];
+		struct i_mmap_shard *tmp_shard = shards[i];
+
+		j = i;
+		while (j > 0 && shard_idx[j - 1] > tmp_idx) {
+			shard_idx[j] = shard_idx[j - 1];
+			shards[j] = shards[j - 1];
+			j--;
+		}
+		shard_idx[j] = tmp_idx;
+		shards[j] = tmp_shard;
+	}
+
+	for (i = 0; i < nr_shards; i++)
+		down_write(&shards[i]->rwsem);
+
+	return nr_shards;
+}
+
+void i_mmap_unlock_shards_write(struct i_mmap_shard *shards[], int nr_shards)
+{
+	int i;
+
+	for (i = nr_shards - 1; i >= 0; i--)
+		up_write(&shards[i]->rwsem);
 }
\ No newline at end of file
diff --git a/mm/mmap.c b/mm/mmap.c
index 0e3461d1e76b..e17e770b540e 100644
--- a/mm/mmap.c
+++ b/mm/mmap.c
@@ -1735,6 +1735,8 @@ struct dup_mmap_file_batch {
 	bool locked;
 	struct vm_area_struct *tmp[8];
 	struct vm_area_struct *prev[8];
+	struct i_mmap_shard *locked_shards[I_MMAP_NR_SHARDS];
+	int nr_locked_shards;
 };
 
 static void dup_mmap_file_batch_flush(struct dup_mmap_file_batch *batch)
@@ -1745,14 +1747,72 @@ static void dup_mmap_file_batch_flush(struct dup_mmap_file_batch *batch)
 	if (batch->count == 0)
 		return;
 
-	flush_dcache_mmap_lock(mapping);
-	for (i = 0; i < batch->count; i++) {
-		if (vma_is_shared_maywrite(batch->tmp[i]))
-			mapping_allow_writable(mapping);
-		vma_interval_tree_insert_after(batch->tmp[i], batch->prev[i],
-					       get_i_mmap_root(mapping));
+	if (mapping_sharded(mapping)) {
+		struct i_mmap_shards *shards_ptr = mapping->i_mmap_shards_ptr;
+		struct i_mmap_shard *shards[I_MMAP_NR_SHARDS];
+		unsigned int shard_idx[I_MMAP_NR_SHARDS];
+		int nr_shards = 0;
+
+		for (i = 0; i < batch->count; i++) {
+			unsigned int idx;
+			int j;
+
+			idx = vma_shard_idx(batch->tmp[i]) % shards_ptr->nr_shards;
+			for (j = 0; j < nr_shards; j++) {
+				if (shard_idx[j] == idx)
+					break;
+			}
+			if (j < nr_shards)
+				continue;
+
+			shard_idx[nr_shards] = idx;
+			shards[nr_shards] = &shards_ptr->shards[idx];
+			nr_shards++;
+		}
+
+		for (i = 1; i < nr_shards; i++) {
+			unsigned int tmp_idx = shard_idx[i];
+			struct i_mmap_shard *tmp_shard = shards[i];
+			int j = i;
+
+			while (j > 0 && shard_idx[j - 1] > tmp_idx) {
+				shard_idx[j] = shard_idx[j - 1];
+				shards[j] = shards[j - 1];
+				j--;
+			}
+			shard_idx[j] = tmp_idx;
+			shards[j] = tmp_shard;
+		}
+
+		for (i = 0; i < nr_shards; i++)
+			down_write(&shards[i]->rwsem);
+
+		flush_dcache_mmap_lock(mapping);
+		for (i = 0; i < batch->count; i++) {
+			unsigned int idx;
+			struct i_mmap_shard *shard;
+
+			if (vma_is_shared_maywrite(batch->tmp[i]))
+				mapping_allow_writable(mapping);
+			idx = vma_shard_idx(batch->tmp[i]) % shards_ptr->nr_shards;
+			shard = &shards_ptr->shards[idx];
+			vma_interval_tree_insert(batch->tmp[i], &shard->tree);
+			atomic_inc(&shards_ptr->vma_count);
+		}
+		flush_dcache_mmap_unlock(mapping);
+
+		for (i = nr_shards - 1; i >= 0; i--)
+			up_write(&shards[i]->rwsem);
+	} else {
+		flush_dcache_mmap_lock(mapping);
+		for (i = 0; i < batch->count; i++) {
+			if (vma_is_shared_maywrite(batch->tmp[i]))
+				mapping_allow_writable(mapping);
+			vma_interval_tree_insert_after(batch->tmp[i], batch->prev[i],
+						       &mapping->i_mmap);
+		}
+		flush_dcache_mmap_unlock(mapping);
 	}
-	flush_dcache_mmap_unlock(mapping);
 	batch->count = 0;
 }
 
@@ -1761,7 +1821,13 @@ static void dup_mmap_file_batch_unlock(struct dup_mmap_file_batch *batch)
 	if (!batch->locked)
 		return;
 	dup_mmap_file_batch_flush(batch);
-	i_mmap_unlock_write(batch->mapping);
+	if (batch->nr_locked_shards > 0) {
+		i_mmap_unlock_shards_write(batch->locked_shards,
+					   batch->nr_locked_shards);
+		batch->nr_locked_shards = 0;
+	} else {
+		i_mmap_unlock_write(batch->mapping);
+	}
 	batch->locked = false;
 	batch->mapping = NULL;
 }
@@ -1773,7 +1839,7 @@ __latent_entropy int dup_mmap(struct mm_struct *mm, struct mm_struct *oldmm)
 	unsigned long charge = 0;
 	LIST_HEAD(uf);
 	VMA_ITERATOR(vmi, mm, 0);
-	struct dup_mmap_file_batch batch = { .count = 0, .mapping = NULL, .locked = false };
+	struct dup_mmap_file_batch batch = { .count = 0, .mapping = NULL, .locked = false, .nr_locked_shards = 0 };
 
 	if (mmap_write_lock_killable(oldmm))
 		return -EINTR;
@@ -1868,7 +1934,11 @@ __latent_entropy int dup_mmap(struct mm_struct *mm, struct mm_struct *oldmm)
 				dup_mmap_file_batch_unlock(&batch);
 			if (!batch.locked) {
 				batch.mapping = mapping;
-				i_mmap_lock_write(mapping);
+				if (mapping_sharded(mapping)) {
+					batch.nr_locked_shards = 0;
+				} else {
+					i_mmap_lock_write(mapping);
+				}
 				batch.locked = true;
 			}
 			batch.tmp[batch.count] = tmp;
diff --git a/mm/vma.c b/mm/vma.c
index 3cf71ad82a3e..ca3cf4ca979d 100644
--- a/mm/vma.c
+++ b/mm/vma.c
@@ -42,6 +42,8 @@ struct mmap_state {
 	bool hold_file_rmap_lock :1;
 	/* If .mmap_prepare changed the file, we don't need to pin. */
 	bool file_doesnt_need_get :1;
+
+	struct i_mmap_shard *held_shard;
 };
 
 #define MMAP_STATE(name, mm_, vmi_, addr_, len_, pgoff_, vm_flags_, file_) \
@@ -231,7 +233,13 @@ static void __vma_link_file(struct vm_area_struct *vma,
 		mapping_allow_writable(mapping);
 
 	flush_dcache_mmap_lock(mapping);
-	vma_interval_tree_insert(vma, get_i_mmap_root(mapping));
+	if (mapping_sharded(mapping)) {
+		struct i_mmap_shard *shard = vma_i_mmap_shard(vma, mapping);
+		vma_interval_tree_insert(vma, &shard->tree);
+		atomic_inc(&mapping->i_mmap_shards_ptr->vma_count);
+	} else {
+		vma_interval_tree_insert(vma, &mapping->i_mmap);
+	}
 	flush_dcache_mmap_unlock(mapping);
 }
 
@@ -245,7 +253,13 @@ static void __remove_shared_vm_struct(struct vm_area_struct *vma,
 		mapping_unmap_writable(mapping);
 
 	flush_dcache_mmap_lock(mapping);
-	vma_interval_tree_remove(vma, get_i_mmap_root(mapping));
+	if (mapping_sharded(mapping)) {
+		struct i_mmap_shard *shard = vma_i_mmap_shard(vma, mapping);
+		vma_interval_tree_remove(vma, &shard->tree);
+		atomic_dec(&mapping->i_mmap_shards_ptr->vma_count);
+	} else {
+		vma_interval_tree_remove(vma, &mapping->i_mmap);
+	}
 	flush_dcache_mmap_unlock(mapping);
 }
 
@@ -294,7 +308,16 @@ static void vma_prepare(struct vma_prepare *vp)
 			uprobe_munmap(vp->adj_next, vp->adj_next->vm_start,
 				      vp->adj_next->vm_end);
 
-		i_mmap_lock_write(vp->mapping);
+		if (mapping_sharded(vp->mapping)) {
+			struct vm_area_struct *lock_vmas[5] = {
+				vp->vma, vp->adj_next, vp->insert,
+				vp->remove, vp->remove2
+			};
+			vp->nr_locked_shards = i_mmap_lock_shards_write(
+				vp->mapping, vp->locked_shards, lock_vmas, 5);
+		} else {
+			i_mmap_lock_write(vp->mapping);
+		}
 		if (vp->insert && vp->insert->vm_file) {
 			/*
 			 * Put into interval tree now, so instantiated pages
@@ -316,10 +339,22 @@ static void vma_prepare(struct vma_prepare *vp)
 
 	if (vp->file) {
 		flush_dcache_mmap_lock(vp->mapping);
-		vma_interval_tree_remove(vp->vma, get_i_mmap_root(vp->mapping));
-		if (vp->adj_next)
-			vma_interval_tree_remove(vp->adj_next,
-						 get_i_mmap_root(vp->mapping));
+		if (mapping_sharded(vp->mapping)) {
+			struct i_mmap_shard *shard;
+
+			shard = vma_i_mmap_shard(vp->vma, vp->mapping);
+			vma_interval_tree_remove(vp->vma, &shard->tree);
+			if (vp->adj_next) {
+				shard = vma_i_mmap_shard(vp->adj_next, vp->mapping);
+				vma_interval_tree_remove(vp->adj_next,
+							 &shard->tree);
+			}
+		} else {
+			vma_interval_tree_remove(vp->vma, &vp->mapping->i_mmap);
+			if (vp->adj_next)
+				vma_interval_tree_remove(vp->adj_next,
+							 &vp->mapping->i_mmap);
+		}
 	}
 
 }
@@ -336,10 +371,21 @@ static void vma_complete(struct vma_prepare *vp, struct vma_iterator *vmi,
 			 struct mm_struct *mm)
 {
 	if (vp->file) {
-		if (vp->adj_next)
-			vma_interval_tree_insert(vp->adj_next,
-						 get_i_mmap_root(vp->mapping));
-		vma_interval_tree_insert(vp->vma, get_i_mmap_root(vp->mapping));
+		if (mapping_sharded(vp->mapping)) {
+			struct i_mmap_shard *shard;
+
+			if (vp->adj_next) {
+				shard = vma_i_mmap_shard(vp->adj_next, vp->mapping);
+				vma_interval_tree_insert(vp->adj_next, &shard->tree);
+			}
+			shard = vma_i_mmap_shard(vp->vma, vp->mapping);
+			vma_interval_tree_insert(vp->vma, &shard->tree);
+		} else {
+			if (vp->adj_next)
+				vma_interval_tree_insert(vp->adj_next,
+							 &vp->mapping->i_mmap);
+			vma_interval_tree_insert(vp->vma, &vp->mapping->i_mmap);
+		}
 		flush_dcache_mmap_unlock(vp->mapping);
 	}
 
@@ -365,7 +411,11 @@ static void vma_complete(struct vma_prepare *vp, struct vma_iterator *vmi,
 	}
 
 	if (vp->file) {
-		i_mmap_unlock_write(vp->mapping);
+		if (mapping_sharded(vp->mapping))
+			i_mmap_unlock_shards_write(vp->locked_shards,
+						   vp->nr_locked_shards);
+		else
+			i_mmap_unlock_write(vp->mapping);
 
 		if (!vp->skip_vma_uprobe) {
 			uprobe_mmap(vp->vma);
@@ -1777,12 +1827,26 @@ static void unlink_file_vma_batch_process(struct unlink_vma_file_batch *vb)
 	int i;
 
 	mapping = vb->vmas[0]->vm_file->f_mapping;
-	i_mmap_lock_write(mapping);
-	for (i = 0; i < vb->count; i++) {
-		VM_WARN_ON_ONCE(vb->vmas[i]->vm_file->f_mapping != mapping);
-		__remove_shared_vm_struct(vb->vmas[i], mapping);
+
+	if (mapping_sharded(mapping)) {
+		struct i_mmap_shard *shards[I_MMAP_NR_SHARDS];
+		int nr_shards;
+
+		nr_shards = i_mmap_lock_shards_write(mapping, shards,
+						      vb->vmas, vb->count);
+		for (i = 0; i < vb->count; i++) {
+			VM_WARN_ON_ONCE(vb->vmas[i]->vm_file->f_mapping != mapping);
+			__remove_shared_vm_struct(vb->vmas[i], mapping);
+		}
+		i_mmap_unlock_shards_write(shards, nr_shards);
+	} else {
+		i_mmap_lock_write(mapping);
+		for (i = 0; i < vb->count; i++) {
+			VM_WARN_ON_ONCE(vb->vmas[i]->vm_file->f_mapping != mapping);
+			__remove_shared_vm_struct(vb->vmas[i], mapping);
+		}
+		i_mmap_unlock_write(mapping);
 	}
-	i_mmap_unlock_write(mapping);
 
 	unlink_file_vma_batch_init(vb);
 }
@@ -1807,17 +1871,29 @@ void unlink_file_vma_batch_final(struct unlink_vma_file_batch *vb)
 		unlink_file_vma_batch_process(vb);
 }
 
-static void vma_link_file(struct vm_area_struct *vma, bool hold_rmap_lock)
+static void vma_link_file(struct vm_area_struct *vma,
+			 struct i_mmap_shard **held_shard)
 {
 	struct file *file = vma->vm_file;
 	struct address_space *mapping;
 
 	if (file) {
 		mapping = file->f_mapping;
-		i_mmap_lock_write(mapping);
-		__vma_link_file(vma, mapping);
-		if (!hold_rmap_lock)
-			i_mmap_unlock_write(mapping);
+		if (mapping_sharded(mapping)) {
+			struct i_mmap_shard *shard = vma_i_mmap_shard(vma, mapping);
+
+			down_write(&shard->rwsem);
+			__vma_link_file(vma, mapping);
+			if (held_shard)
+				*held_shard = shard;
+			else
+				up_write(&shard->rwsem);
+		} else {
+			i_mmap_lock_write(mapping);
+			__vma_link_file(vma, mapping);
+			if (!held_shard)
+				i_mmap_unlock_write(mapping);
+		}
 	}
 }
 
@@ -1831,7 +1907,7 @@ static int vma_link(struct mm_struct *mm, struct vm_area_struct *vma)
 
 	vma_start_write(vma);
 	vma_iter_store_new(&vmi, vma);
-	vma_link_file(vma, /* hold_rmap_lock= */false);
+	vma_link_file(vma, NULL);
 	mm->map_count++;
 	validate_mm(mm);
 	return 0;
@@ -2552,7 +2628,7 @@ static int __mmap_new_vma(struct mmap_state *map, struct vm_area_struct **vmap)
 	vma_start_write(vma);
 	vma_iter_store_new(vmi, vma);
 	map->mm->map_count++;
-	vma_link_file(vma, map->hold_file_rmap_lock);
+	vma_link_file(vma, map->hold_file_rmap_lock ? &map->held_shard : NULL);
 
 	/*
 	 * vma_merge_new_range() calls khugepaged_enter_vma() too, the below
@@ -2708,11 +2784,11 @@ static int call_action_complete(struct mmap_state *map,
 
 	ret = mmap_action_complete(action, vma);
 
-	/* If we held the file rmap we need to release it. */
 	if (map->hold_file_rmap_lock) {
-		struct file *file = vma->vm_file;
-
-		i_mmap_unlock_write(file->f_mapping);
+		if (map->held_shard)
+			up_write(&map->held_shard->rwsem);
+		else
+			i_mmap_unlock_write(vma->vm_file->f_mapping);
 	}
 	return ret;
 }
diff --git a/mm/vma.h b/mm/vma.h
index eba388c61ef4..9f9d2aec0de2 100644
--- a/mm/vma.h
+++ b/mm/vma.h
@@ -7,6 +7,8 @@
 #ifndef __MM_VMA_H
 #define __MM_VMA_H
 
+struct i_mmap_shard;
+
 /*
  * VMA lock generalization
  */
@@ -21,6 +23,9 @@ struct vma_prepare {
 	struct vm_area_struct *remove2;
 
 	bool skip_vma_uprobe :1;
+
+	struct i_mmap_shard *locked_shards[5];
+	int nr_locked_shards;
 };
 
 struct unlink_vma_file_batch {
-- 
2.43.0

